psi4/src/psi4/libmints/angularmomentum.cc

```diff
@@ -26,242 +26,108 @@
-#include "psi4/libmints/molecule.h"
 #include "psi4/libmints/integral.h"
+#include "mcmurchiedavidson.h"
 
 #include <memory>
 #include <stdexcept>
 
 #include <libint2/shell.h>
 
 using namespace psi;
+using namespace mdintegrals;
 
-// Initialize overlap_recur_ to +1 basis set angular momentum, +1 on each center is sufficient
-// to compute the dipole derivatives
 AngularMomentumInt::AngularMomentumInt(std::vector<SphericalTransform>& spherical_transforms,
                                        std::shared_ptr<BasisSet> bs1, std::shared_ptr<BasisSet> bs2, int nderiv)
-    : OneBodyAOInt(spherical_transforms, bs1, bs2, nderiv), overlap_recur_(bs1->max_am() + 1, bs2->max_am() + 1) {
-    int maxam1 = bs1_->max_am();
-    int maxam2 = bs2_->max_am();
+    : OneBodyAOInt(spherical_transforms, bs1, bs2, nderiv), MDHelper(bs1->max_am(), bs2->max_am()) {
+    int maxnao1 = INT_NCART(maxam1_);
+    int maxnao2 = INT_NCART(maxam2_);
 
-    int maxnao1 = (maxam1 + 1) * (maxam1 + 2) / 2;
-    int maxnao2 = (maxam2 + 1) * (maxam2 + 2) / 2;
-
-    // Increase buffer size to handle x, y, and z components
     if (deriv_ == 0) {
+        // one chunk for each Cartesian component
         buffer_ = new double[3 * maxnao1 * maxnao2];
         set_chunks(3);
     } else {
-        throw PSIEXCEPTION("AngularMomentumInt does not provide derivatives");
+        throw PSIEXCEPTION("AngularMomentumInt does not provide derivatives.");
     }
     buffers_.resize(nchunk_);
 }
 
 AngularMomentumInt::~AngularMomentumInt() { delete[] buffer_; }
 
-
-void AngularMomentumInt::compute_pair(const libint2::Shell &s1, const libint2::Shell &s2) {
-    int ao12;
+void AngularMomentumInt::compute_pair(const libint2::Shell& s1, const libint2::Shell& s2) {
+    // Computes the angular momentum integrals for a pair of shells using eq 9.3.33
+    // equation numbers from Molecular Electronic-Structure Theory (10.1002/9781119019572)
     int am1 = s1.contr[0].l;
     int am2 = s2.contr[0].l;
-    int nprim1 = s1.nprim();
-    int nprim2 = s2.nprim();
-    double A[3], B[3];
-    double Sxy1, Sxy2, Sxz1, Sxz2;
-    double Syx1, Syx2, Syz1, Syz2;
-    double Szx1, Szx2, Szy1, Szy2;
-    double S0x1, S0x2, S0y1, S0y2, S0z1, S0z2;
-    double muxy1, muxy2, muxz1, muxz2;
-    double muyx1, muyx2, muyz1, muyz2;
-    double muzx1, muzx2, muzy1, muzy2;
-
-    A[0] = s1.O[0];
-    A[1] = s1.O[1];
-    A[2] = s1.O[2];
-    B[0] = s2.O[0];
-    B[1] = s2.O[1];
-    B[2] = s2.O[2];
+    int am = am1 + am2;
 
-    int ydisp = INT_NCART(am1) * INT_NCART(am2);
-    int zdisp = ydisp + INT_NCART(am1) * INT_NCART(am2);
+    const auto& comps_am1 = am_comps_[am1];
+    const auto& comps_am2 = am_comps_[am2];
 
-    // compute intermediates
-    double AB2 = 0.0;
-    AB2 += (A[0] - B[0]) * (A[0] - B[0]);
-    AB2 += (A[1] - B[1]) * (A[1] - B[1]);
-    AB2 += (A[2] - B[2]) * (A[2] - B[2]);
+    auto A = s1.O;
+    auto B = s2.O;
+    // TODO: settable origin? (not supported previously)
```

Comment:
I don't see a use case for a non-zero origin, but I think we should allow all integrals involving a third center to have that be settable.  I think that is more consistent behavior between the different engines, and it does not affect any of the code that does the computation.  Right now the Potential base class handles the setting of the origin - perhaps we should move that to the OneBody base class, and have the derived classes that don't support an origin (like kinetic, overlap) throw an error.  If we go this route, it's probably a separate PR because it will touch quite a few different objects in a very minor way.

Reply:
Yes -- `set_origin` already exists in `OneBody`, but making this consistent among as many integral classes as possible is probably beyond the scope of the PR, especially since the "feature" is not needed right now.